gradio/routes.py

```diff
@@ -400,33 +395,32 @@ async def reset_iterator(body: ResetBody):
                 app.iterators_to_reset[body.session_hash].add(body.fn_index)
             return {"success": True}
 
-        async def run_predict(
+        # had to use '/run' endpoint for Colab compatibility, '/api' supported for backwards compatibility
+        @app.post("/run/{api_name}", dependencies=[Depends(login_check)])
+        @app.post("/run/{api_name}/", dependencies=[Depends(login_check)])
+        @app.post("/api/{api_name}", dependencies=[Depends(login_check)])
+        @app.post("/api/{api_name}/", dependencies=[Depends(login_check)])
+        async def predict(
+            api_name: str,
             body: PredictBody,
-            request: Request | List[Request],
-            fn_index_inferred: int,
+            request: fastapi.Request,
+            username: str = Depends(get_current_user),
         ):
+            fn_index_inferred = route_utils.infer_fn_index(
+                app=app, api_name=api_name, body=body
+            )
+
+            gr_request = route_utils.compile_gr_request(
```

Comment:
Very clever @whitphx ! The one thing that jumped out at me is that this breaks `api_open=False`. Some app developers don't want requests to skip the queue so I don't think we should break that functionality.

Reply:
Thank you, and sorry that's right, I will revert it.

Reply:
Do you think this change on the if block makes sense?

Before:
```python
            if (
                not app.get_blocks().api_open
                and app.get_blocks().queue_enabled_for_fn(fn_index_inferred)
                and f"Bearer {app.queue_token}" != request.headers.get("Authorization")
            ):
                raise HTTPException(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    detail="Not authorized to skip the queue",
                )
```

After:
```python
            if not app.get_blocks().api_open:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                )
```


https://github.com/gradio-app/gradio/pull/5165/files#diff-324a7165f5d5a8823a28b76f5653fa45f32c8144c82b2e528882c97c7eae534fR409-R412

Reply:
I think we still need to check if the queue is enabled for the function. Otherwise I think this would work. Also, I think 401 code is still appropriate here, no?

Reply:
> I think we still need to check if the queue is enabled for the function.

Ahh, thank you, I misunderstood this if-condition. This API endpoint can be closed only when the queue is enabled as another gateway to access the functionality.

>  Also, I think 401 code is still appropriate here, no?

Actually I don't have a solid opinion on it and 401 would be ok.

Just let me dump my thoughts:
I thought 401 would implies that there is still possibility for the client to make auth successful by providing the correct credentials, and the failure is due to client's responsibility. However, now the availability of this endpoint is controlled completely by the server-side configs (queue and `api_open` flag), so 404 may make more sense.